src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14392,217 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
+	}
+
+	if (!g_strstrip (direct_pinvoke)) {
+		res = DIRECT_PINVOKE_SKIP;
+		goto cleanup;
+	}
+
+	char **direct_pinvoke_split = g_strsplit (direct_pinvoke, "!", 2);
+	if (!direct_pinvoke_split) {
+		aot_printerrf (acfg, "Failed to split the provided 'direct_pinvoke' AOT option '%s' with delimiter '!'\n", direct_pinvoke);
+		res = DIRECT_PINVOKE_ERR;
+		goto cleanup;
+	}
+
+	*module_name_ptr = g_strdup (direct_pinvoke_split[0]);
+	if (!*module_name_ptr) {
+		aot_printerrf (acfg, "Failed to strdup the module name portion of direct pinvoke '%s'.\n", direct_pinvoke);
+		res = DIRECT_PINVOKE_ERR;
+	}
+	*entrypoint_name_ptr = g_strdup (direct_pinvoke_split[1]);
+	if (direct_pinvoke_split[1] && !*entrypoint_name_ptr) {
+		aot_printerrf (acfg, "Failed to strdup the entrypoint name portion of direct pinvoke '%s'.\n", direct_pinvoke);
+		res = DIRECT_PINVOKE_ERR;
+	}
+	g_strfreev (direct_pinvoke_split);
+	direct_pinvoke_split = NULL;
+
+cleanup:
+	g_free (direct_pinvoke);
+	direct_pinvoke = NULL;
+
+early_exit:
+	return res;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// add_direct_pinvoke adds the module and entrypoint of a specified direct pinvoke to the
+// MonoAotCompile instance's HashTable of module/entrypoint entries.
+// It is presumed that module_name_ptr and entrypoint_name_ptr point to valid strings.
+// It transfers ownership of the module_name and entrypoint_name strings to the HashTable
+//
+// Arguments:
+//  * acfg - the MonoAotCompiler instance
+//  ** module_name_ptr - the pointer to the module name (assumed not NULL)
+//  ** entrypoint_name_ptr - the pointer to the entrypoint name
+//
+// Return Value:
+//  gboolean corresponding to whether or not the direct_pinvoke was successfully added
+//  or if an error occurred.
+//
+
+static gboolean
+add_direct_pinvoke (MonoAotCompile *acfg, char **module_name_ptr, char **entrypoint_name_ptr)
+{
+	gboolean success = TRUE;
+	// MODULE
+	// All entrypoints from the library are direct
+	if (!*entrypoint_name_ptr) {
+		g_hash_table_insert (acfg->direct_pinvokes, *module_name_ptr, NULL);
+		goto cleanup;
+	}
+
+	// MODULE!ENTRYPOINT
+	GHashTable *entrypoints;
+	if (g_hash_table_lookup_extended (acfg->direct_pinvokes, *module_name_ptr, NULL, (gpointer *)&entrypoints)) {
+		// All entrypoints from the library are direct
+		if (!entrypoints)
+			goto cleanup;
+
+		g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
+	} else {
+		entrypoints = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+		if (!entrypoints) {
+			aot_printerrf (acfg, "Failed to allocate new entrypoints HashTable.\n");
+			success = FALSE;
+			goto cleanup;
+		}
+		g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
+		g_hash_table_insert (acfg->direct_pinvokes, *module_name_ptr, entrypoints);
+	}
+
+	entrypoints = NULL;
+cleanup:
+	g_free (*module_name_ptr);
+	g_free (*entrypoint_name_ptr);
+	*module_name_ptr, module_name_ptr = NULL;
```

Comment:
You only need to set *module_name_ptr and *entrypoint_name_ptr to NULL.

Reply:
Ah, I think I had the misconception that there was an address block that stored the `&module` (address of `module` (the address block that stores the address of the string data)) passed into this function, and I thought that we would need to clear the value. Instead, is what happens in the local scope of this function that a copy of the value of `&module` passed in, and upon leaving the scope of the function, whatever is used to store that memory is freed?

Reply:
What matters is setting the *module_name_ptr, *entrypoint_name_ptr to NULL since that is what affects the callers pointer (you pass it "by reference"), just setting module_name_ptr, entrypoint_name_ptr  will only set the local pointer to NULL (the reference but not what it points to).